libfwupdplugin/fu-intel-nvm.c

```diff
@@ -0,0 +1,880 @@
+		guint8 available_sections = 0;
+
+		if (!fu_intel_nvm_read_uint8(self,
+					     FU_INTEL_NVM_SECTION_DIGITAL,
+					     FU_INTEL_NVM_DIGITAL_OFFSET_AVAILABLE_SECTIONS,
+					     &available_sections,
+					     error)) {
+			g_prefix_error(error, "failed to read available sections: ");
+			return FALSE;
+		}
+		if (!fu_intel_nvm_read_uint16(self,
+					      FU_INTEL_NVM_SECTION_DIGITAL,
+					      FU_INTEL_NVM_DIGITAL_OFFSET_UCODE,
+					      &ucode_offset,
+					      error)) {
+			g_prefix_error(error, "failed to read ucode offset: ");
+			return FALSE;
+		}
+		offset = ucode_offset;
+		if ((available_sections & DRAM_FLAG) == 0) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INVALID_FILE,
+					    "cannot find needed FW sections in the FW image file");
+			return FALSE;
+		}
+
+		for (guint8 i = 1; i < DRAM_FLAG; i <<= 1) {
+			if (available_sections & i) {
+				if (!fu_intel_nvm_read_ucode_section_len(self,
+									 offset,
+									 &ucode_offset,
+									 error))
+					return FALSE;
+				offset += ucode_offset;
+			}
+		}
+		priv->sections[FU_INTEL_NVM_SECTION_DRAM_UCODE] =
+		    offset + priv->sections[FU_INTEL_NVM_SECTION_DIGITAL];
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_intel_nvm_missing_needed_drom(FuIntelNvm *self)
+{
+	FuIntelNvmPrivate *priv = GET_PRIVATE(self);
+	if (priv->sections[FU_INTEL_NVM_SECTION_DROM] != 0)
+		return FALSE;
+	if (priv->is_host && priv->gen < 3)
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_intel_nvm_parse(FuFirmware *firmware,
+		   GBytes *fw,
+		   gsize offset,
+		   FwupdInstallFlags flags,
+		   GError **error)
+{
+	FuIntelNvm *self = FU_INTEL_NVM(firmware);
+	FuIntelNvmPrivate *priv = GET_PRIVATE(self);
+	guint8 tmp = 0;
+	guint16 version_raw = 0;
+	struct {
+		guint16 device_id;
+		guint gen;
+		FuIntelNvmFamily family;
+		guint ports;
+	} hw_info_arr[] = {{0x156D, 2, FU_INTEL_NVM_FAMILY_FR, 2},   /* FR 4C */
+			   {0x156B, 2, FU_INTEL_NVM_FAMILY_FR, 1},   /* FR 2C */
+			   {0x157E, 2, FU_INTEL_NVM_FAMILY_WR, 1},   /* WR */
+			   {0x1578, 3, FU_INTEL_NVM_FAMILY_AR, 2},   /* AR 4C */
+			   {0x1576, 3, FU_INTEL_NVM_FAMILY_AR, 1},   /* AR 2C */
+			   {0x15C0, 3, FU_INTEL_NVM_FAMILY_AR, 1},   /* AR LP */
+			   {0x15D3, 3, FU_INTEL_NVM_FAMILY_AR_C, 2}, /* AR-C 4C */
+			   {0x15DA, 3, FU_INTEL_NVM_FAMILY_AR_C, 1}, /* AR-C 2C */
+			   {0x15E7, 3, FU_INTEL_NVM_FAMILY_TR, 1},   /* TR 2C */
+			   {0x15EA, 3, FU_INTEL_NVM_FAMILY_TR, 2},   /* TR 4C */
+			   {0x15EF, 3, FU_INTEL_NVM_FAMILY_TR, 2},   /* TR 4C device */
+			   {0x15EE, 3, FU_INTEL_NVM_FAMILY_BB, 0},   /* BB device */
+			   /* Maple ridge devices
+			    * NOTE: These are expected to be flashed via UEFI capsules *not*
+			    * Thunderbolt plugin Flashing via fwupd will require matching kernel
+			    * work. They're left here only for parsing the binaries
+			    */
+			   {0x1136, 4, FU_INTEL_NVM_FAMILY_MR, 2},
+			   {0x1137, 4, FU_INTEL_NVM_FAMILY_MR, 2},
+			   {0}};
```

Comment:
probably missing entries for Goshen Ridge here

Reply:
Ya, I pinged @bleungatchromium for those